### PR TITLE
Added in basic support for broadcast nested loop join

### PR DESCRIPTION
**Pull request by revans2**
I would appreciate some reviews on this.

It is a part of #265 but is missing `CartesianExec` which is an implementation that shows up if one of the tables is too larger to be broadcast and it is an `inner` or `cross` join with no equality comparison.

It adds in support for Cross equality joins, that in those cases are the same as an Inner join.

It also add in support for BroadcastNestedLoopJoin on Cross and Inner joins.  The biggest issue is the amount of memory that could be used by a Cross join this big.

I plan on trying to use the current memory size of each table (left and right) to decide if we should play some games with memory.  If the size is too large then we break the tables down into smaller pieces.

i.e.

```
left_size_per_row = left_size_memory/left_rows
right_size_per_row = right_size_memory/right_rows

if (((left_size_per_row + right_size_per_row) * left_rows * right_rows) > target_batch_size) {
  split the tables.  Preferable just split the stream table, but if we cannot, then split both and loop.
}
```

This would still not fix all cases, because we could broadcast something really large and blow up from just trying to hold it in memory.

### CODE REVIEW

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  override protected def doExecute(): RDD[InternalRow] =
+    throw new IllegalStateException("This should only be called from columnar")
+
+  override lazy val additionalMetrics: Map[String, SQLMetric] = Map(
+    "buildDataSize" -> SQLMetrics.createSizeMetric(sparkContext, "build side size"),
+    "buildTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "build time"),
+    "joinTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "join time"),
+    "joinOutputRows" -> SQLMetrics.createMetric(sparkContext, "join output rows"),
+    "filterTime" -> SQLMetrics.createNanoTimingMetric(sparkContext, "filter time"))
+  //TODO need to figure out what if anything we can make common with GpuHashJoin
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastNestedLoopJoinExec.scala:102]: Still planning to do this, but I wanted to figure out the split logic first, if we think it is needed.

**Comment by jlowe** [comment]
Took a quick glance, seems fine to me.  As you noted, would be great to explore sharing a lot of the boilerplate build type handling, output distribution, etc. that is common with the existing hash join.

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
I updated the code so `BroadcastnestedLoopJoin` is off by default.  I could not find a clean way to reuse anything between it and the other join implementations.  It would take about as much code to make it common as it saved, so I have left them separate.  We might revisit it again in the future if someone has a better idea on how to do it (I tried a trait to mix in).

I also filed a follow on issue #302 to try and fix some of the memory issues and let us turn this on by default.

**Pull request merged**